File: sorts/merge_sort/python/merge_sort.py
def merge_sort(data):
    if len(data) <= 1:
        return data

    midle = len(data) / 2
    left = merge_sort(data[:midle])
    right = merge_sort(data[midle:])
    return merge(left, right)


def merge(left, right):
    res = []

    while left and right:
        if left[0] < right[0]:
            res.append(left.pop(0))
        else:
            res.append(right.pop(0))

    if left:
        res.extend(left)
    if right:
        res.extend(right)

    return res

if __name__ == '__main__':
    data = [1, 3, 5, 7, 8]
    data1 = [2, 4, 6]
    print merge(data, data1)
    d = [8, 3, 6, 1, 90, 12, 3]
    print merge_sort(d)

File: sorts/merge_sort/go/merge_sort.go
package main

import (
	"fmt"
)

func main() {
	fmt.Println("sdfsdf")
	data := []int{1, 3, 5}
	data1 := []int{2, 4, 6}
	fmt.Println(merge(data, data1))
	data = []int{10, 2, 4, 3, 90, 23, 2, 7}
	fmt.Println(merge_sort(data))
}

func merge_sort(data []int) []int {
	if len(data) <= 1 {
		return data
	}

	middle := len(data) / 2

	left := merge_sort(data[middle:])
	right := merge_sort(data[:middle])

	return merge(left, right)
}

func merge(left []int, right []int) []int {
	res := make([]int, len(left)+len(right))
	i := 0
	j := 0

	for i < len(left) && j < len(right) {
		if left[i] < right[j] {
			res[i+j] = left[i]
			i++
		} else {
			res[i+j] = right[j]
			j++
		}
	}

	for i < len(left) {
		res[i+j] = left[i]
		i++
	}

	for j < len(right) {
		res[i+j] = right[j]
		j++
	}

	return res
}
